File: 电话号码的字母组合.cpp
class Solution {
public:
    vector<string> letterCombinations(string digits) {
        int n = digits.size();
        string mp[10] = {"", "", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"};
        vector<string> ans;
        if (digits == "") {
            return {};
        }
        function<void(int, string)> dfs = [&](int i, string s) {
            if (i == n) {
                ans.emplace_back(s);
                return;
            }
            for (char c : mp[digits[i] - '0']) {
                s[i] = c;
                dfs(i + 1, s);
            }
        };
        dfs(0, string(n, '0'));
        return ans;
    }
};

File: 电话号码的字母组合.py
class Solution:
    def letterCombinations(self, digits: str) -> List[str]:
        n = len(digits)
        mp = ["", "", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"]
        ans = []
        if digits == "":
            return []
        
        def dfs(i, s):
            if i == n:
                ans.append("".join(s))
                return
            for c in list(mp[int(digits[i])]):
                s[i] = c
                dfs(i + 1, s)
        
        dfs(0, ['0' for _ in range(n)])
        return ans
